### PR TITLE
OpenID Connect client management

### PR DESCRIPTION
**Pull request by byewokko**
API for the registration and management of OpenID Connect clients.

# API

## `POST /client`

- **Client registration** according to [OpenID Connect Dynamic Client Registration 1.0](https://openid.net/specs/openid-connect-registration-1_0.html)
- JSON body with Client metadata (as described in https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata)
- Response contains `client_id`, `client_secret` and optionally other stuff (as described in https://openid.net/specs/openid-connect-registration-1_0.html#RegistrationResponse)

##  `POST /client/{client_id}/reset_secret`

- **Get a new client secret**

## `PUT /client/{client_id}`

- **Edit client metadata**
- Same body as in client registration

## `GET /client/{client_id}`

- **Get client detail**

## `GET /client`

- **List registered clients**

# Provisioning

Since SeaCat web UI also requires a client ID registered with its respective redirect URIs, it is necessary to register it during provisioning. The client parameters can be set in the `provisioning.json` config:

```json
{
 "webui_clients": [
  {
   "client_id": "seacat-webui",
   "redirect_uris": ["https://auth.test.loc/seacat"],
  }
 ]
}
```

### CODE REVIEW

File: seacatauth/client/service.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+			resources.append(resource_dict)
+
+		return {
+			"data": resources,
+			"count": count,
+		}
+
+
+	async def get(self, client_id: str):
+		data = await self.StorageService.get(self.ClientCollection, client_id)
```
Comment (ateska) [inline seacatauth/client/service.py:56]: Be careful not to send Secret data out ...

File: seacatauth/client/service.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+
+	async def create(
+		self,
+		client_id: str,
+		base_url: str,
+		description: str = None,
+		scope: list = None,
+	):
+		# TODO: Generate client secret
```
Comment (ateska) [inline seacatauth/client/service.py:67]: Yes, it should be generated and shown to the user exactly once, as a result of this call.

File: seacatauth/client/service.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+					"Client ID must consist only of characters 'a-z0-9._-', "
+					"start with a letter, and be between 3 and 32 characters long.",
+			}
+
+		upsertor = self.StorageService.upsertor(self.ClientCollection, obj_id=client_id)
+
+		upsertor.set("bu", base_url)  # TODO: Validate
+
+		client_secret = secrets.token_urlsafe(self.ClientSecretLength)
+		upsertor.set("cs", client_secret)
```
Comment (ateska) [inline seacatauth/client/service.py:82]: `upsertor.set("_cs", client_secret)` ... it is a secret

File: seacatauth/client/service.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+	async def update(
+		self,
+		client_id: str,
+		base_url: str = None,
+		description: str = None,
+		scope: list = None,
+	):
+		raise NotImplementedError()
```
Comment (ateska) [inline seacatauth/client/service.py:116]: Please add some kind of comment why it is not implemented and what is the plan.

File: seacatauth/client/service.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+
+	async def delete(self, client_id: str):
+		raise NotImplementedError()
```
Comment (ateska) [inline seacatauth/client/service.py:120]: dtto ^

**Review by ateska — CHANGES_REQUESTED** [review]

File: seacatauth/client/service.py (hunk 1 above)
Comment (byewokko) [reply seacatauth/client/service.py]: Client secret was removed from the GET calls

File: seacatauth/client/service.py (hunk 2 above)
Comment (byewokko) [reply seacatauth/client/service.py]: Implemented

File: seacatauth/openidconnect/client/handler.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+class OAuth2ClientHandler(object):
+	def __init__(self, app, client_svc):
+		self.ClientService = client_svc
+
+		web_app = app.WebContainer.WebApp
+		web_app.router.add_get("/openidconnect/client", self.list)
+		web_app.router.add_get("/openidconnect/client/{client_id}", self.get)
+		web_app.router.add_post("/openidconnect/client", self.create)
+		web_app.router.add_post("/openidconnect/client/{client_id}/reset_secret", self.reset_secret)
+		web_app.router.add_put("/openidconnect/client/{client_id}", self.update)
```
Comment (ateska) [inline seacatauth/openidconnect/client/handler.py:26]: It should not be part of "openidconnect" - hence the public API

File: seacatauth/openidconnect/client/service.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+from .. import exceptions
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class OAuth2ClientService(asab.Service):
```
Comment (ateska) [inline seacatauth/openidconnect/client/service.py:18]: A small comment about the reason for this service would be great.

**Review by ateska — CHANGES_REQUESTED** [review]

File: seacatauth/openidconnect/client/service.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+	# 	"^client_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# 	"^policy_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# 	"^tos_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# }
+}
+
+
+class ClientService(asab.Service):
+	"""
+	Provides API for the registration and management of OpenID Connect clients
```
Comment (ateska) [inline seacatauth/openidconnect/client/service.py:92]: Please specify the name of the standard which this service implements.
Together with a canonical URL to the specifications.

File: seacatauth/openidconnect/client/service.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+		return client
+
+
+	async def rest_get(self, client_id: str):
+		client = await self._get(client_id)
+		client = self._rest_normalize(client, include_client_secret=True)
+		return client
+
+
+	def _rest_normalize(self, client: dict, include_client_secret: bool = False):
```
Comment (ateska) [inline seacatauth/openidconnect/client/service.py:147]: REST normalization is a handler task, not a service.

**Review by ateska — CHANGES_REQUESTED** [review]

**Pull request merged**